### PR TITLE
Use a BMP decoder when resizing an image

### PR DESCRIPTION
**Pull request by calixteman**
The image decoding won't block the main thread any more.
For now, it isn't enabled for Chrome because issue6741.pdf leads to a crash.

### CODE REVIEW

File: src/shared/util.js [hunk 1]
```diff
@@ -634,19 +634,22 @@ class FeatureTest {
       (typeof navigator !== "undefined" &&
         typeof navigator?.platform === "string")
     ) {
+      const isFirefox =
+        (typeof PDFJSDev !== "undefined" && PDFJSDev.test("MOZCENTRAL")) ||
+        (typeof navigator?.userAgent === "string" &&
+          navigator.userAgent.includes("Firefox"));
       return shadow(this, "platform", {
+        isChrome: !isFirefox && !!window.chrome,
```
Comment (Snuffleupagus) [inline src/shared/util.js:642]: Unfortunately this isn't safe, since code in `src/shared/` will run on both the main- and worker-threads (and in Node.js) so you cannot access `window` here.

Can we please just revert the changes in this file (I'll make a few other suggestions)?

File: web/app.js [hunk 2]
```diff
@@ -1045,6 +1045,7 @@ const PDFViewerApplication = {
     const loadingTask = getDocument({
       ...apiParams,
       ...args,
+      isChrome: FeatureTest.platform.isChrome,
```
Comment (Snuffleupagus) [inline web/app.js:1048]: Please revert this.
```suggestion
```

File: src/display/api.js [hunk 3]
```diff
@@ -177,6 +177,7 @@ const DefaultStandardFontDataFactory =
  *   `OffscreenCanvas` in the worker. Primarily used to improve performance of
  *   image conversion/rendering.
  *   The default value is `true` in web environments and `false` in Node.js.
+ * @property {boolean} [isChrome] - Determines if we can use bmp ImageDecoder.
```
Comment (Snuffleupagus) [inline src/display/api.js:180]:
```suggestion
 * @property {boolean} [isChrome] - Determines if we can use bmp ImageDecoder.
 *   NOTE: Temporary option until [link to Chromium bug report] is fixed.
```

File: src/display/api.js [hunk 4]
```diff
@@ -281,6 +282,7 @@ function getDocument(src = {}) {
     typeof src.isOffscreenCanvasSupported === "boolean"
       ? src.isOffscreenCanvasSupported
       : !isNodeJS;
+  const isChrome = src.isChrome === true;
```
Comment (Snuffleupagus) [inline src/display/api.js:285]: To avoid issues, and to support this properly when running tests, can we please change it as follows?
```suggestion
  const isChrome =
    // eslint-disable-next-line no-nested-ternary
    typeof src.isChrome === "boolean"
      ? src.isChrome
      : (typeof PDFJSDev === "undefined" || !PDFJSDev.test("MOZCENTRAL")) &&
        !FeatureTest.platform.isFirefox &&
        typeof window !== "undefined" &&
        !!window?.chrome;
```

File: src/core/image_resizer.js [hunk 5]
```diff
@@ -160,10 +167,33 @@ class ImageResizer {
...
     const data = this._encodeBMP();
-    const blob = new Blob([data.buffer], {
-      type: "image/bmp",
-    });
-    const bitmapPromise = createImageBitmap(blob);
+
+    let decoder, imagePromise;
+
+    // TODO: remove the isChrome, once Chrome isn't crashing anymore with
+    // isssue6741.pdf.
```
Comment (Snuffleupagus) [inline src/core/image_resizer.js:174]:
```suggestion
    // issue6741.pdf.
```

File: src/core/image_resizer.js [hunk 6]
```diff
@@ -160,10 +167,33 @@ class ImageResizer {
...
+    if (
+      (typeof PDFJSDev !== "undefined" && PDFJSDev.test("MOZCENTRAL")) ||
```
Comment (Snuffleupagus) [inline src/core/image_resizer.js:176]: Unfortunately I don't believe that enabling this unconditionally in Firefox is safe, since this can be disabled with a preference *and* that preference is only enabled by default in Nightly: https://searchfox.org/mozilla-central/rev/ec342a3d481d9ac3324d1041e05eefa6b61392d2/modules/libpref/init/StaticPrefList.yaml#3241-3245

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Do you have benchmarking results that show (clear) improvements with this patch?

Have you reported the Chrome-crash to the browser developers?

**Comment by calixteman** [comment]
> Do you have benchmarking results that show (clear) improvements with this patch?

With the `issue16263.pdf`, I made 15 measures in using `performance.now()` and getting the image from the bmp buffer takes:
 - ~2.4s (stddev ~0.6s) with `ImageDecoder`
 - ~3.3s (stddev ~1.1s) with `createImageBitmap`

with a debug build of Firefox on Windows 11.
As far as I know, it shouldn't make that much difference in term of decoding because at the end the same code with the same data is called on the C++ side, but with `ImageDecoder` we can avoid a copy of the buffer thanks to the `transfer` parameter. The main difference is that the decoding happens in an other thread than the main content thread. I think that decoding on the main thread explains the higher variability (see the standard deviation in the `createImageBitmap` case).
From my point of view the main interest of this patch is to not have the decoding blocking the main thread and as a side effect if we can win few tens of millis it isn't that bad either.

> Have you reported the Chrome-crash to the browser developers?

Yep we reached the Chrome dev in charge of that stuff yesterday and they filed a bug for the crash:
https://issues.chromium.org/issues/374807001

File: src/shared/util.js
Comment (Snuffleupagus) [inline src/shared/util.js:1]: Please revert the changes in this file, since they're no longer needed.

File: src/core/image_resizer.js [hunk 7]
```diff
@@ -88,6 +90,21 @@ class ImageResizer {
...
+    return shadow(
+      this,
+      "canUseImageDecoder",
+      // eslint-disable-next-line no-undef
+      this.#isChrome || typeof ImageDecoder === "undefined"
+        ? Promise.resolve(false)
+        : // eslint-disable-next-line no-undef
+          ImageDecoder.isTypeSupported("image/bmp")
+    );
+  }
```
Comment (Snuffleupagus) [inline src/core/image_resizer.js:106]: Nit: Can we please move this up, such that this getter isn't placed "between" the closely related `MAX_DIM` and `MAX_AREA` getters?

File: src/core/image_resizer.js [hunk 8]
```diff
@@ -188,7 +224,13 @@ class ImageResizer {
 
     let newWidth = width;
     let newHeight = height;
-    let bitmap = await bitmapPromise;
+    let bitmap;
+    if (decoder) {
+      ({ image: bitmap } = await imagePromise);
+      decoder.close();
```
Comment (Snuffleupagus) [inline src/core/image_resizer.js:230]: Do we need to handle this being a rejected Promise, if decoding fails?

**Comment by calixteman** [comment]
The bug is fixed in Chrome, I checked in  Canary and everything is working correctly.
/botio test

File: src/core/evaluator.js [hunk 9]
```diff
@@ -232,7 +233,10 @@ class PartialEvaluator {
 
     this._regionalImageCache = new RegionalImageCache();
     this._fetchBuiltInCMapBound = this.fetchBuiltInCMap.bind(this);
-    ImageResizer.setMaxArea(this.options.canvasMaxAreaInBytes);
+    ImageResizer.setOptions({
+      isChrome: this.options.isChrome,
+      maxArea: this.options.canvasMaxAreaInBytes,
+    });
```
Comment (Snuffleupagus) [inline src/core/evaluator.js:239]: We could reduce the impact of the new option a little bit in the Firefox PDF Viewer.
```suggestion
    if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("MOZCENTRAL")) {
      ImageResizer.setMaxArea(this.options.canvasMaxAreaInBytes);
    } else {
      ImageResizer.setOptions({
        isChrome: this.options.isChrome,
        maxArea: this.options.canvasMaxAreaInBytes,
      });
    }
```

File: src/core/image_resizer.js [hunk 10]
```diff
@@ -113,6 +130,11 @@ class ImageResizer {
     }
   }
 
+  static setOptions({ maxArea = -1, isChrome = false }) {
+    this.setMaxArea(maxArea);
+    this.#isChrome = isChrome;
+  }
```
Comment (Snuffleupagus) [inline src/core/image_resizer.js:136]: Similar to the previous comment.
```suggestion
  static setOptions(opts) {
    if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("MOZCENTRAL")) {
      throw new Error("Not implemented: setOptions");
    }
    this.setMaxArea(opts.maxArea ?? -1);
    this.#isChrome = opts.isChrome ?? false;
  }
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, with a couple of suggestions to further reduce the impact on the Firefox PDF Viewer; thank you.
